src/cephadm/cephadm

```diff
@@ -5212,6 +5257,462 @@ def command_gather_facts():
+</body>
+</html>""".format(api_version=CephadmDaemonHandler.api_version)
+
+    def _fetch_root(self):
+        self.send_response(200)
+        self.send_header('Content-type', 'text/html; charset=utf-8')
+        self.end_headers()
+        self.wfile.write(self._help_page().encode('utf-8'))
+
+    def do_GET(self):
+        """Handle *all* GET requests"""
+
+        # we could wrap this to include some form of auth, but at this point the 
+        # data is read only and we offer no means of remotely invoking a cephadm
+        # command - so auth is not required (we're basically in the same category
+        # as the prometheus node-exporter)
+        if self.path == '/':
+            # provide a html response if someone hits the root url, to document the 
+            # available api endpoints
+            return self._fetch_root()
+        elif self.path in CephadmDaemonHandler.valid_routes:
+            u = self.path.split('/')[-1]
+            data = json.dumps({})
+            if u == 'metadata':
+                data = json.dumps(self.server.cephadm_cache)
+            elif u == 'daemons':
+                data = json.dumps(self.server.cephadm_cache.get('daemons', {}))
+            elif u == 'disks':
+                data = json.dumps(self.server.cephadm_cache.get('disks', {}))    
+            elif u == 'health':
+                data = json.dumps(self.server.cephadm_cache.get('health', {}))
+            elif u == 'host':
+                data = json.dumps(self.server.cephadm_cache.get('host', {}))
+            
+            self.send_response(200)
+            self.send_header('Content-type','application/json')
+            self.end_headers()
+            self.wfile.write(data.encode('utf-8'))
+        else:
+            # Invalid GET URL
+            bad_request_msg = "Valid URLs are: {}".format(', '.join(CephadmDaemonHandler.valid_routes))
+            self.send_response(404, message=bad_request_msg)  # reason
+            self.send_header('Content-type','application/json')
+            self.end_headers()
+            self.wfile.write(json.dumps({"message": bad_request_msg}).encode('utf-8'))
+    
+    def log_message(self, format, *args):
+        rqst = " ".join(str(a) for a in args)
+        logger.info(f"client:{self.address_string()} [{self.log_date_time_string()}] {rqst}")
+
+
+class CephadmHTTPServer(ThreadingMixIn, HTTPServer):
+    allow_reuse_address = True
+    daemon_threads = True
+    cephadm_cache = None
+
+
+class CephadmDaemon():
+
+    daemon_type = "cephadm"
+    default_port = 5003
+    bin_name = 'cephadm'
+    loop_delay = 1
+    thread_check_interval = 5
+
+    def __init__(self, fsid, daemon_id, port=None):
+        self.fsid = fsid
+        self.daemon_id = daemon_id 
+        if not port:
+            self.port = CephadmDaemon.default_port
+        else:
+            self.port = port
+        self.workers = []
+        self.http_server = None
+        self.stop = False
+        self.cephadm_cache = {
```

Comment:
This dict is used a lot!!. To include type annotations could help us to avoid problems in the future

Reply:
dict replaced by an object that handles update and locking